mmdet3d/core/bbox/coders/monoflex_bbox_coder.py

```diff
@@ -0,0 +1,453 @@
+
+        # 2 dimensions for projected centers2d offsets
+        pred_offsets2d = bbox[:, 4:6]
+
+        # 3 dimensions for 3D bbox dimensions offsets
+        pred_dimensions_offsets3d = bbox[:, 29:32]
+
+        # the first 8 dimensions are for orientation bin classification
+        # and the second 8 dimensions are for orientation offsets.
+        pred_orientations = torch.cat((bbox[:, 32:40], bbox[:, 40:48]), dim=1)
+
+        # 3 dimensions for the uncertainties of the solved depths from
+        # groups of keypoints
+        pred_keypoints_depth_uncertainty = bbox[:, 26:29]
+
+        # 1 dimension for the uncertainty of directly regressed depth
+        pred_direct_depth_uncertainty = bbox[:, 49:50].squeeze(-1)
+
+        # 2 dimension of offsets x keypoints (8 corners + top/bottom center)
+        pred_keypoints2d = bbox[:, 6:26]
+
+        # 1 dimension for depth offsets
+        pred_depth_offsets = bbox[:, 48:49].squeeze(-1)
+
+        # decode the pred residual dimensions to real dimensions
+        pred_dimensions = self.decode_dimension(labels,
+                                                pred_dimensions_offsets3d)
+        pred_direct_depth = self.decode_direct_depth(pred_depth_offsets)
+        pred_keypoints_depth = self.decode_depth_from_keypoints(
+            pred_keypoints2d, pred_dimensions, cam2imgs, downsample_ratio)
+
+        pred_direct_depth_uncertainty = torch.clamp(
+            pred_direct_depth_uncertainty, self.uncertainty_range[0],
+            self.uncertainty_range[1])
+        pred_keypoints_depth_uncertainty = torch.clamp(
+            pred_keypoints_depth_uncertainty, self.uncertainty_range[0],
+            self.uncertainty_range[1])
+
+        if self.use_combined_depth:
+            pred_combined_uncertainty = torch.cat(
+                (pred_direct_depth_uncertainty.unsqueeze(-1),
+                 pred_keypoints_depth_uncertainty),
+                dim=1).exp()
+            pred_combined_depth = torch.cat(
+                (pred_direct_depth.unsqueeze(-1), pred_keypoints_depth), dim=1)
+            pred_uncertainty_weights = 1 / pred_combined_uncertainty
+            pred_uncertainty_weights = \
+                pred_uncertainty_weights / \
+                pred_uncertainty_weights.sum(dim=1, keepdim=True)
+            pred_combined_depth = torch.sum(
+                pred_combined_depth * pred_uncertainty_weights, dim=1)
+        else:
+            pred_combined_depth = None
+
+        preds = dict(
+            bboxes2d=pred_bboxes2d,
+            dimensions=pred_dimensions,
+            offsets2d=pred_offsets2d,
+            direct_depth=pred_direct_depth,
+            keypoints2d=pred_keypoints2d,
+            keypoints_depth=pred_keypoints_depth,
+            combined_depth=pred_combined_depth,
+            orientations=pred_orientations,
+        )
+
+        return preds
+
+    def decode_direct_depth(self, depth_offsets):
```

Comment:
what is the difference between direct depth and depth?

Reply:
in monoflex depth should includes directly regressed depth and keypoints depth.